src/parser/queries/QueryBuilder.ts

```diff
@@ -26,50 +34,114 @@ export class QueryBuilder {
         this.statements = [];
     }
 
-    build(): Query {
+    /**
+     * Builds the query for finding the currently set statements in code of the currently set language
+     * @return The query, or undefined if the file extension / language of the source file is not supported.
+     */
+    build(): Query | undefined {
         let queryString = "";
+        if (this.language !== Languages.Unsupported) {
+            if (this.statements.length > 0) {
+                queryString = this.getStatementsQuery();
+            }
+
+            dlog("------------- Start Query: --------------");
+            dlog(queryString);
+            dlog("-----------------------------------------");
 
-        if (this.statements.length > 0) {
-            queryString = this.getBruteForcedStatementsQuery();
+            return new Query(this.treeSitterLanguage, queryString);
         }
+    }
 
-        dlog("------------- Start Query: --------------");
-        dlog(queryString);
-        dlog("-----------------------------------------");
+    /**
+     * Generates a query string for finding the currently set statements in code of the currently set language.
+     * Returns an empty query if the language is not supported.
+     * @return A query string.
+     * @private
+     */
+    private getStatementsQuery() {
+        const statementQueries: string[] = [];
 
-        return new Query(this.treeSitterLanguage, queryString);
+        for (const statement of this.statements) {
+            // Is "activatedFor" still necessary? I have only found one node with the activated_for_languages filed.
+            // However, it seems helpful for some (e.g. debugging) purposes,
+            // so I kept it and introduced "applicableFor" along it.
+            if (statement.applicableFor(this.language) && statement.activatedFor(this.language)) {
+                statementQueries.push(statement.toQuery());
+            }
+        }
+        return statementQueries.join("\n");
     }
 
-    // This is not necessary anymore
-    // due to the mapped node types (/resources/node-types-mapped.config)
-    // and the information per expressions
-    // in which language it is included
-    // Thus, we do not have to figure out
-    // if the expression is included in a language by query the expression
-    // and catch an exception if it is not included.
```

Comment:
In general, this change implements the removal of the bruteforce approach already planned by @ce-bo as stated in this comment.